Flaky AS NATS test

<!--
Thanks for submitting a bug report. Please fill the template below,
otherwise we will not be able to process this bug report.
-->

#### Summary
<!-- Summarize the problem in a few sentences: -->

Flaky AS NATS test

#### Steps to Reproduce
<!-- How can we reproduce the problem? -->

Random

<!--
Please upload relevant configuration (as .txt).
If you use the command "ttn-lw-stack config", you can redact sensitive config.
-->

#### What do you see now?
<!--
Please paste terminal output, upload logs (as .txt) or upload screenshots.
-->

```
  WARN No cookie hash key configured, generated a random one hash_key=CA90E1F03FA690E96351D8FB1163251BE7205552608AD070C6297727C322A7C100EF2DCD28DDF676B5B4B4529159585A748B4CC14F01F2685C4277897AF3E535 namespace=web test_name=TestApplicationServer
  WARN No cookie block key configured, generated a random one block_key=B2E6930B986F933271CB083C3912A2A2B056AE91196DAB0FA69E46DED7F24E39 namespace=web test_name=TestApplicationServer
  WARN No static assets found in any search path namespace=web search_paths=[] test_name=TestApplicationServer
  WARN No rights TTL configured                 test_name=TestApplicationServer
  WARN No cluster key configured, generated a random one key=58faddf0ecadac5f9f637c0e088435f5f8466f7fc1a0a89edabac396815ebd7c test_name=TestApplicationServer
  INFO Listening for connections                address=:9184 namespace=grpc protocol=gRPC test_name=TestApplicationServer
  INFO Listening for connections                address=:8099 namespace=web protocol=Web test_name=TestApplicationServer
  INFO Linked                                   application_uid=foo-app namespace=applicationserver network_server=ns test_name=TestApplicationServer
  INFO Subscribed                               application_uid=foo-app grpc_method=Subscribe grpc_service=ttn.lorawan.v3.AppAs namespace=applicationserver/io/grpc remote_addr=pipe request_id=01DG4K7VXZEKX3GKJ8D3AH9Q14 test_name=TestApplicationServer
  INFO Unsubscribed                             application_uid=foo-app grpc_method=Subscribe grpc_service=ttn.lorawan.v3.AppAs namespace=applicationserver/io/grpc remote_addr=pipe request_id=01DG4K7VXZEKX3GKJ8D3AH9Q14 test_name=TestApplicationServer
  WARN Finished streaming call                  duration=2.048457872s error=context canceled grpc_method=Subscribe grpc_service=ttn.lorawan.v3.AppAs namespace=grpc request_id=01DG4K7VXZEKX3GKJ8D3AH9Q14 test_name=TestApplicationServer
  WARN Finished streaming call                  duration=932.957µs error=error:pkg/auth/rights:no_application_rights (no rights for application `foo-app`) error_correlation_id=a5a973c3896743c9bb59988d03de8fab error_name=no_application_rights error_namespace=pkg/auth/rights grpc_code=PermissionDenied grpc_method=Subscribe grpc_service=ttn.lorawan.v3.AppAs namespace=grpc request_id=01DG4K7XY0S0Z15Z5CRB8GFYXF test_name=TestApplicationServer
  WARN Finished streaming call                  duration=541.678µs error=error:pkg/auth/rights:no_application_rights (no rights for application `invalid-application`) error_correlation_id=210204d477db4da882c6f747891e3f87 error_name=no_application_rights error_namespace=pkg/auth/rights grpc_code=PermissionDenied grpc_method=Subscribe grpc_service=ttn.lorawan.v3.AppAs namespace=grpc request_id=01DG4K7XY11FX41KG7PCHMSYME test_name=TestApplicationServer
  INFO Subscribed                               application_uid=foo-app grpc_method=Subscribe grpc_service=ttn.lorawan.v3.AppAs namespace=applicationserver/io/grpc remote_addr=pipe request_id=01DG4K7XY2QBQ42837ZSTQFNRA test_name=TestApplicationServer
  WARN Drop downlink message; session not found or AppSKey not available application_uid=foo-app count=3 device_uid=foo-app.foo-device f_cnt=12 f_port=12 namespace=applicationserver network_server=ns new_session_key_id=44 next_a_f_cnt_down=85 session_key_id=11223344 test_name=TestApplicationServer
  WARN Failed to process upstream message       application_uid=foo-app error=error:pkg/applicationserver:device_not_found (device `foo-app.bar-device` not found) namespace=applicationserver network_server=ns test_name=TestApplicationServer
  WARN Failed to process upstream message       application_uid=foo-app error=error:pkg/applicationserver:device_not_found (device `foo-app.bar-device` not found) namespace=applicationserver network_server=ns test_name=TestApplicationServer
  INFO Finished unary call                      duration=1.064318ms error=error:pkg/applicationserver:device_not_found (device `foo-app.bar-device` not found) error_correlation_id=584f9b7bc98f4325b9846e4e36a33918 error_name=device_not_found error_namespace=pkg/applicationserver grpc_code=NotFound grpc_method=DownlinkQueuePush grpc_service=ttn.lorawan.v3.AppAs namespace=grpc request_id=01DG4K87ZGJS24S3G86BA4K032 test_name=TestApplicationServer
  INFO Finished unary call                      duration=2.152703ms grpc_method=DownlinkQueuePush grpc_service=ttn.lorawan.v3.AppAs namespace=grpc request_id=01DG4K89ZHJVW15MYJ2CEES28C test_name=TestApplicationServer
  INFO Finished unary call                      duration=2.944889ms grpc_method=DownlinkQueuePush grpc_service=ttn.lorawan.v3.AppAs namespace=grpc request_id=01DG4K8BZHJW3QHJ8K8QPRDRH0 test_name=TestApplicationServer
  INFO Finished unary call                      duration=1.453864ms grpc_method=DownlinkQueueReplace grpc_service=ttn.lorawan.v3.AppAs namespace=grpc request_id=01DG4K8DZJ87CJ6T5KTHJ6PX0R test_name=TestApplicationServer
  INFO Unsubscribed                             application_uid=foo-app grpc_method=Subscribe grpc_service=ttn.lorawan.v3.AppAs namespace=applicationserver/io/grpc remote_addr=pipe request_id=01DG4K7XY2QBQ42837ZSTQFNRA test_name=TestApplicationServer
  WARN Finished streaming call                  duration=18.48105067s error=context canceled grpc_method=Subscribe grpc_service=ttn.lorawan.v3.AppAs namespace=grpc request_id=01DG4K7XY2QBQ42837ZSTQFNRA test_name=TestApplicationServer
  INFO Connected                                application_uid=foo-app namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35342 test_name=TestApplicationServer
  INFO Disconnected                             application_uid=foo-app namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35342 test_name=TestApplicationServer
  WARN Failed to setup connection               error=error:pkg/auth/rights:no_application_rights (no rights for application `foo-app`) namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35346 test_name=TestApplicationServer
  WARN Failed to setup connection               error=error:pkg/auth/rights:no_application_rights (no rights for application `foo-app`) namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35348 test_name=TestApplicationServer
  WARN Failed to setup connection               error=error:pkg/auth/rights:no_application_rights (no rights for application `invalid-application`) namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35352 test_name=TestApplicationServer
  WARN Failed to setup connection               error=error:pkg/auth/rights:no_application_rights (no rights for application `invalid-application`) namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35354 test_name=TestApplicationServer
  INFO Connected                                application_uid=foo-app namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35358 test_name=TestApplicationServer
  WARN Drop downlink message; session not found or AppSKey not available application_uid=foo-app count=3 device_uid=foo-app.foo-device f_cnt=12 f_port=12 namespace=applicationserver network_server=ns new_session_key_id=44 next_a_f_cnt_down=85 session_key_id=11223344 test_name=TestApplicationServer
  WARN Failed to process upstream message       application_uid=foo-app error=error:pkg/applicationserver:device_not_found (device `foo-app.bar-device` not found) namespace=applicationserver network_server=ns test_name=TestApplicationServer
  WARN Failed to process upstream message       application_uid=foo-app error=error:pkg/applicationserver:device_not_found (device `foo-app.bar-device` not found) namespace=applicationserver network_server=ns test_name=TestApplicationServer
  WARN Failed to handle downlink messages       application_uid=foo-app error=error:pkg/applicationserver:device_not_found (device `foo-app.bar-device` not found) namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35358 test_name=TestApplicationServer topic=v3/foo-app/devices/bar-device/down/push
  INFO Disconnected                             application_uid=foo-app namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35358 test_name=TestApplicationServer
  INFO Finished unary call                      duration=1.400271ms grpc_method=Set grpc_service=ttn.lorawan.v3.ApplicationPubSubRegistry namespace=grpc request_id=01DG4K98147PW65SEEHSF5JF21 test_name=TestApplicationServer
  INFO Started                                  application_uid=foo-app namespace=applicationserver/io/pubsub provider=*ttnpb.ApplicationPubSub_NATS pub_sub_id=test test_name=TestApplicationServer
  WARN Finished unary call                      duration=751.208µs error=error:pkg/auth/rights:no_application_rights (no rights for application `foo-app`) error_correlation_id=b626925a3ee64fce8c140b233406371f error_name=no_application_rights error_namespace=pkg/auth/rights grpc_code=PermissionDenied grpc_method=Set grpc_service=ttn.lorawan.v3.ApplicationPubSubRegistry namespace=grpc request_id=01DG4K9A1444G0TT28MHP8WFBR test_name=TestApplicationServer
  WARN Finished unary call                      duration=505.353µs error=error:pkg/auth/rights:no_application_rights (no rights for application `foo-app`) error_correlation_id=894aaf0d399a49be8345bb48362e332f error_name=no_application_rights error_namespace=pkg/auth/rights grpc_code=PermissionDenied grpc_method=Set grpc_service=ttn.lorawan.v3.ApplicationPubSubRegistry namespace=grpc request_id=01DG4K9A164E0JST222EMR2A7Y test_name=TestApplicationServer
  WARN Failed to receive downlink queue operation application_uid=foo-app error=context canceled namespace=applicationserver/io/pubsub operation=push provider=*ttnpb.ApplicationPubSub_NATS pub_sub_id=test test_name=TestApplicationServer
  INFO Finished unary call                      duration=1.562081ms grpc_method=Set grpc_service=ttn.lorawan.v3.ApplicationPubSubRegistry namespace=grpc request_id=01DG4K9A17JAWTQEA2AEDZ18QX test_name=TestApplicationServer
  WARN Failed to receive downlink queue operation application_uid=foo-app error=context canceled namespace=applicationserver/io/pubsub operation=replace provider=*ttnpb.ApplicationPubSub_NATS pub_sub_id=test test_name=TestApplicationServer
  INFO Integration canceled                     application_uid=foo-app namespace=applicationserver/io/pubsub provider=*ttnpb.ApplicationPubSub_NATS pub_sub_id=test test_name=TestApplicationServer
  INFO Unlinked                                 application_uid=foo-app namespace=applicationserver network_server=ns test_name=TestApplicationServer
--- FAIL: TestApplicationServer (55.51s)
    --- FAIL: TestApplicationServer/Traffic/pubsub/nats (4.10s)
        --- FAIL: TestApplicationServer/Traffic/pubsub/nats/Upstream (2.05s)
            --- FAIL: TestApplicationServer/Traffic/pubsub/nats/Upstream/RegisteredDevice/JoinAccept (2.05s)
                applicationserver_test.go:1504: Expected upstream timeout
FAIL
FAIL	go.thethings.network/lorawan-stack/pkg/applicationserver	55.576s
```

#### What do you want to see instead?
<!-- Please add some examples or mock-ups if applicable. -->

No errors

#### Environment
<!--
Your environment: OS/Browser/Gateway/Device/...? Versions? IDs/EUIs?
Paste the output of "ttn-lw-cli version" or "ttn-lw-stack version" if applicable.
-->

Travis, i386

#### How do you propose to implement this?
<!-- Please think about how this could be fixed. -->

Probably increase timeouts. AS is already taking a lot of time though.

#### Can you do this yourself and submit a Pull Request?
<!-- You can also @mention experts if you need help with this. -->

Can review

--- a/config/messages.json
+++ b/config/messages.json
@@ -1601,6 +1601,15 @@
       "file": "format.go"
     }
   },
+  "error:pkg/applicationserver/io/pubsub:integration_failed": {
+    "translations": {
+      "en": "integration {pub_sub_id} failed"
+    },
+    "description": {
+      "package": "pkg/applicationserver/io/pubsub",
+      "file": "observability.go"
+    }
+  },
   "error:pkg/applicationserver/io/web/redis:invalid_fieldmask": {
     "translations": {
       "en": "invalid fieldmask"
@@ -5954,6 +5963,51 @@
     },
     "description": {
       "package": "pkg/applicationserver",
+      "file": "observability.go"
+    }
+  },
+  "event:as.pubsub.delete": {
+    "translations": {
+      "en": "delete pubsub"
+    },
+    "description": {
+      "package": "pkg/applicationserver/io/pubsub",
+      "file": "observability.go"
+    }
+  },
+  "event:as.pubsub.fail": {
+    "translations": {
+      "en": "fail pubsub"
+    },
+    "description": {
+      "package": "pkg/applicationserver/io/pubsub",
+      "file": "observability.go"
+    }
+  },
+  "event:as.pubsub.set": {
+    "translations": {
+      "en": "set pubsub"
+    },
+    "description": {
+      "package": "pkg/applicationserver/io/pubsub",
+      "file": "observability.go"
+    }
+  },
+  "event:as.pubsub.start": {
+    "translations": {
+      "en": "start pubsub"
+    },
+    "description": {
+      "package": "pkg/applicationserver/io/pubsub",
+      "file": "observability.go"
+    }
+  },
+  "event:as.pubsub.stop": {
+    "translations": {
+      "en": "stop pubsub"
+    },
+    "description": {
+      "package": "pkg/applicationserver/io/pubsub",
       "file": "observability.go"
     }
   },

--- a/pkg/applicationserver/applicationserver_test.go
+++ b/pkg/applicationserver/applicationserver_test.go
@@ -40,6 +40,7 @@ import (
 	"go.thethings.network/lorawan-stack/pkg/component"
 	"go.thethings.network/lorawan-stack/pkg/config"
 	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/events"
 	"go.thethings.network/lorawan-stack/pkg/jsonpb"
 	"go.thethings.network/lorawan-stack/pkg/rpcmetadata"
 	"go.thethings.network/lorawan-stack/pkg/ttnpb"
@@ -426,6 +427,8 @@ hardware_versions:
 				return ids == registeredApplicationID && key == registeredApplicationKey
 			},
 			Connect: func(ctx context.Context, t *testing.T, ids ttnpb.ApplicationIdentifiers, key string, chs *connChannels) error {
+				evCh := make(chan events.Event, EventsBufferSize)
+				defer test.RedirectEvents(evCh)()
 				// Configure pubsub.
 				creds := grpc.PerRPCCredentials(rpcmetadata.MD{
 					AuthType:      "Bearer",
@@ -495,6 +498,9 @@ hardware_versions:
 				if _, err := client.Set(ctx, req, creds); err != nil {
 					return err
 				}
+				if !test.WaitEvent(ctx, evCh, "as.pubsub.start") {
+					t.Fatal("Integration start timeout")
+				}
 				nc, err := nats_client.Connect("nats://localhost:4124")
 				if err != nil {
 					return err
@@ -555,6 +561,8 @@ hardware_versions:
 				return ids == registeredApplicationID && key == registeredApplicationKey
 			},
 			Connect: func(ctx context.Context, t *testing.T, ids ttnpb.ApplicationIdentifiers, key string, chs *connChannels) error {
+				evCh := make(chan events.Event, EventsBufferSize)
+				defer test.RedirectEvents(evCh)()
 				// Configure pubsub.
 				creds := grpc.PerRPCCredentials(rpcmetadata.MD{
 					AuthType:      "Bearer",
@@ -625,6 +633,9 @@ hardware_versions:
 				}
 				if _, err := client.Set(ctx, req, creds); err != nil {
 					return err
+				}
+				if !test.WaitEvent(ctx, evCh, "as.pubsub.start") {
+					t.Fatal("Integration start timeout")
 				}
 				clientOpts := mqtt.NewClientOptions()
 				clientOpts.AddBroker(mqttLis.Addr().String())

--- a/pkg/applicationserver/applicationserver_util_test.go
+++ b/pkg/applicationserver/applicationserver_util_test.go
@@ -22,7 +22,8 @@ import (
 )
 
 var (
-	Timeout = (1 << 9) * test.Delay
+	Timeout          = (1 << 8) * test.Delay
+	EventsBufferSize = (1 << 6)
 )
 
 // MockDeviceRegistry is a mock DeviceRegistry used for testing.

--- a/pkg/applicationserver/applicationserver_util_test.go
+++ b/pkg/applicationserver/applicationserver_util_test.go
@@ -22,7 +22,7 @@ import (
 )
 
 var (
-	Timeout          = (1 << 8) * test.Delay
+	Timeout          = (1 << 6) * test.Delay
 	EventsBufferSize = (1 << 6)
 )
 

--- a/pkg/applicationserver/io/pubsub/grpc_pubsub.go
+++ b/pkg/applicationserver/io/pubsub/grpc_pubsub.go
@@ -20,6 +20,7 @@ import (
 	pbtypes "github.com/gogo/protobuf/types"
 	"go.thethings.network/lorawan-stack/pkg/auth/rights"
 	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/events"
 	"go.thethings.network/lorawan-stack/pkg/log"
 	"go.thethings.network/lorawan-stack/pkg/ttnpb"
 	"go.thethings.network/lorawan-stack/pkg/unique"
@@ -98,7 +99,7 @@ func (ps *PubSub) Set(ctx context.Context, req *ttnpb.SetApplicationPubSubReques
 		)).WithError(err).Warn("Failed to cancel integration")
 	}
 	ps.startTask(ps.ctx, req.ApplicationPubSubIdentifiers)
-
+	events.Publish(evtSetPubSub(ctx, req.ApplicationPubSubIdentifiers, nil))
 	return pubsub, nil
 }
 
@@ -125,5 +126,6 @@ func (ps *PubSub) Delete(ctx context.Context, ids *ttnpb.ApplicationPubSubIdenti
 	if err != nil {
 		return nil, err
 	}
+	events.Publish(evtDeletePubSub(ctx, *ids, nil))
 	return ttnpb.Empty, nil
 }

--- a/pkg/applicationserver/io/pubsub/observability.go
+++ b/pkg/applicationserver/io/pubsub/observability.go
@@ -1,0 +1,132 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pubsub
+
+import (
+	"context"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/events"
+	"go.thethings.network/lorawan-stack/pkg/metrics"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/unique"
+)
+
+var (
+	evtSetPubSub = events.Define(
+		"as.pubsub.set", "set pubsub",
+		ttnpb.RIGHT_APPLICATION_SETTINGS_BASIC,
+	)
+	evtDeletePubSub = events.Define(
+		"as.pubsub.delete", "delete pubsub",
+		ttnpb.RIGHT_APPLICATION_SETTINGS_BASIC,
+	)
+	evtPubSubStart = events.Define(
+		"as.pubsub.start", "start pubsub",
+		ttnpb.RIGHT_APPLICATION_SETTINGS_BASIC,
+		ttnpb.RIGHT_APPLICATION_TRAFFIC_READ,
+		ttnpb.RIGHT_APPLICATION_TRAFFIC_DOWN_WRITE,
+	)
+	evtPubSubStop = events.Define(
+		"as.pubsub.stop", "stop pubsub",
+		ttnpb.RIGHT_APPLICATION_SETTINGS_BASIC,
+		ttnpb.RIGHT_APPLICATION_TRAFFIC_READ,
+		ttnpb.RIGHT_APPLICATION_TRAFFIC_DOWN_WRITE,
+	)
+	evtPubSubFail = events.Define(
+		"as.pubsub.fail", "fail pubsub",
+		ttnpb.RIGHT_APPLICATION_SETTINGS_BASIC,
+		ttnpb.RIGHT_APPLICATION_TRAFFIC_READ,
+		ttnpb.RIGHT_APPLICATION_TRAFFIC_DOWN_WRITE,
+	)
+)
+
+const (
+	subsystem     = "as_pubsub"
+	unknown       = "unknown"
+	applicationID = "application_id"
+)
+
+var pubsubMetrics = &integrationsMetrics{
+	integrationsStarted: metrics.NewContextualCounterVec(
+		prometheus.CounterOpts{
+			Subsystem: subsystem,
+			Name:      "integrations_started_total",
+			Help:      "Number of integrations started",
+		},
+		[]string{applicationID},
+	),
+	integrationsStopped: metrics.NewContextualCounterVec(
+		prometheus.CounterOpts{
+			Subsystem: subsystem,
+			Name:      "integrations_stopped_total",
+			Help:      "Number of integrations stopped",
+		},
+		[]string{applicationID},
+	),
+	integrationsFailed: metrics.NewContextualCounterVec(
+		prometheus.CounterOpts{
+			Subsystem: subsystem,
+			Name:      "integrations_failed_total",
+			Help:      "Number of integrations failed",
+		},
+		[]string{applicationID},
+	),
+}
+
+func init() {
+	metrics.MustRegister(pubsubMetrics)
+}
+
+type integrationsMetrics struct {
+	integrationsStarted *metrics.ContextualCounterVec
+	integrationsStopped *metrics.ContextualCounterVec
+	integrationsFailed  *metrics.ContextualCounterVec
+}
+
+func (m integrationsMetrics) Describe(ch chan<- *prometheus.Desc) {
+	m.integrationsStarted.Describe(ch)
+	m.integrationsStopped.Describe(ch)
+	m.integrationsFailed.Describe(ch)
+}
+
+func (m integrationsMetrics) Collect(ch chan<- prometheus.Metric) {
+	m.integrationsStarted.Collect(ch)
+	m.integrationsStopped.Collect(ch)
+	m.integrationsFailed.Collect(ch)
+}
+
+func registerIntegrationStart(ctx context.Context, i *integration) {
+	events.Publish(evtPubSubStart(ctx, i.ApplicationIdentifiers, i.ApplicationPubSubIdentifiers))
+	pubsubMetrics.integrationsStarted.WithLabelValues(ctx, i.ApplicationID).Inc()
+}
+
+func registerIntegrationStop(ctx context.Context, i *integration) {
+	events.Publish(evtPubSubStop(ctx, i.ApplicationIdentifiers, i.ApplicationPubSubIdentifiers))
+	pubsubMetrics.integrationsStopped.WithLabelValues(ctx, i.ApplicationID).Inc()
+}
+
+var errIntegrationFailed = errors.DefineAborted("integration_failed", "integration {pub_sub_id} failed")
+
+func registerIntegrationFail(ctx context.Context, i *integration, err error) {
+	err = errIntegrationFailed.
+		WithAttributes(
+			"application_uid", unique.ID(ctx, i.ApplicationIdentifiers),
+			"pub_sub_id", i.PubSubID).
+		WithCause(err)
+	events.Publish(evtPubSubFail(ctx, i.ApplicationIdentifiers, err))
+	pubsubMetrics.integrationsFailed.WithLabelValues(ctx, i.ApplicationID).Inc()
+}

--- a/pkg/applicationserver/io/pubsub/pubsub.go
+++ b/pkg/applicationserver/io/pubsub/pubsub.go
@@ -105,6 +105,7 @@ type integration struct {
 	ttnpb.ApplicationPubSub
 	ctx    context.Context
 	cancel errorcontext.CancelFunc
+	closed chan struct{}
 
 	conn *provider.Connection
 
@@ -236,18 +237,22 @@ func (ps *PubSub) start(ctx context.Context, pb *ttnpb.ApplicationPubSub) (err e
 		ApplicationPubSub: *pb,
 		ctx:               ctx,
 		cancel:            cancel,
+		closed:            make(chan struct{}),
 		server:            ps.server,
 	}
 	if _, loaded := ps.integrations.LoadOrStore(psUID, i); loaded {
+		log.FromContext(ctx).Warn("Integration already started")
 		return errAlreadyConfigured.WithAttributes("application_uid", appUID, "pub_sub_id", pb.PubSubID)
 	}
 	go func() {
 		<-ctx.Done()
 		ps.integrationErrors.Store(psUID, ctx.Err())
 		ps.integrations.Delete(psUID)
-		if err := ctx.Err(); !errors.IsCanceled(err) {
+		if err := ctx.Err(); err != nil && !errors.IsCanceled(err) {
+			log.FromContext(ctx).WithError(err).Warn("Integration failed")
 			registerIntegrationFail(ctx, i, err)
 		}
+		close(i.closed)
 	}()
 	provider, err := provider.GetProvider(pb)
 	if err != nil {
@@ -291,6 +296,7 @@ func (ps *PubSub) stop(ctx context.Context, ids ttnpb.ApplicationPubSubIdentifie
 			"pub_sub_id", ids.PubSubID,
 		)).Debug("Integration canceled")
 		i.cancel(context.Canceled)
+		<-i.closed
 	} else {
 		ps.integrationErrors.Delete(psUID)
 	}

--- a/pkg/applicationserver/io/pubsub/pubsub.go
+++ b/pkg/applicationserver/io/pubsub/pubsub.go
@@ -245,6 +245,9 @@ func (ps *PubSub) start(ctx context.Context, pb *ttnpb.ApplicationPubSub) (err e
 		<-ctx.Done()
 		ps.integrationErrors.Store(psUID, ctx.Err())
 		ps.integrations.Delete(psUID)
+		if err := ctx.Err(); !errors.IsCanceled(err) {
+			registerIntegrationFail(ctx, i, err)
+		}
 	}()
 	provider, err := provider.GetProvider(pb)
 	if err != nil {
@@ -268,10 +271,12 @@ func (ps *PubSub) start(ctx context.Context, pb *ttnpb.ApplicationPubSub) (err e
 	go i.handleUp(ctx)
 	i.startHandleDown(ctx)
 	logger.Info("Started")
+	registerIntegrationStart(ctx, i)
 	<-ctx.Done()
 	if err := ctx.Err(); errors.IsCanceled(err) {
 		logger.Info("Integration canceled")
 		i.conn.Shutdown(ctx)
+		registerIntegrationStop(ctx, i)
 	}
 	return
 }

--- a/pkg/applicationserver/linking.go
+++ b/pkg/applicationserver/linking.go
@@ -88,6 +88,7 @@ type link struct {
 	ttnpb.ApplicationLink
 	ctx    context.Context
 	cancel errorcontext.CancelFunc
+	closed chan struct{}
 
 	conn      *grpc.ClientConn
 	connName  string
@@ -161,12 +162,14 @@ func (as *ApplicationServer) link(ctx context.Context, ids ttnpb.ApplicationIden
 		ApplicationLink:        *target,
 		ctx:                    ctx,
 		cancel:                 cancel,
+		closed:                 make(chan struct{}),
 		connReady:              make(chan struct{}),
 		subscribeCh:            make(chan *io.Subscription, 1),
 		unsubscribeCh:          make(chan *io.Subscription, 1),
 		upCh:                   make(chan *io.ContextualApplicationUp, linkBufferSize),
 	}
 	if _, loaded := as.links.LoadOrStore(uid, l); loaded {
+		log.FromContext(ctx).Warn("Link already started")
 		return errAlreadyLinked.WithAttributes("application_uid", uid)
 	}
 	go func() {
@@ -177,6 +180,7 @@ func (as *ApplicationServer) link(ctx context.Context, ids ttnpb.ApplicationIden
 			log.FromContext(ctx).WithError(err).Warn("Link failed")
 			registerLinkFail(ctx, l, err)
 		}
+		close(l.closed)
 	}()
 	if err := as.connectLink(ctx, l); err != nil {
 		return err
@@ -264,6 +268,7 @@ func (as *ApplicationServer) cancelLink(ctx context.Context, ids ttnpb.Applicati
 		l := val.(*link)
 		log.FromContext(ctx).WithField("application_uid", uid).Debug("Unlink")
 		l.cancel(context.Canceled)
+		<-l.closed
 	} else {
 		as.linkErrors.Delete(uid)
 	}

--- a/pkg/applicationserver/observability.go
+++ b/pkg/applicationserver/observability.go
@@ -253,6 +253,7 @@ func registerLinkStop(ctx context.Context, link *link) {
 
 func registerLinkFail(ctx context.Context, link *link, err error) {
 	events.Publish(evtLinkFail(ctx, link.ApplicationIdentifiers, err))
+	asMetrics.linksFailed.WithLabelValues(ctx, link.NetworkServerAddress).Inc()
 }
 
 func registerSubscribe(ctx context.Context, sub *io.Subscription) {

--- a/pkg/util/test/events.go
+++ b/pkg/util/test/events.go
@@ -156,3 +156,30 @@ func CollectEvents(f func()) []events.Event {
 	f()
 	return evs
 }
+
+// RedirectEvents redirects the published events to the
+// provided channel until the returned function is called.
+func RedirectEvents(ch chan events.Event) func() {
+	return SetDefaultEventsPubSub(&MockEventPubSub{
+		PublishFunc:     func(ev events.Event) { ch <- ev },
+		SubscribeFunc:   func(name string, hdl events.Handler) error { return nil },
+		UnsubscribeFunc: func(name string, hdl events.Handler) {},
+	})
+}
+
+// WaitEvent waits for a specific event to be sent to the channel.
+func WaitEvent(ctx context.Context, ch chan events.Event, name string) bool {
+	for {
+		select {
+		case <-ctx.Done():
+			return false
+		case ev, ok := <-ch:
+			if !ok {
+				panic("channel is closed")
+			}
+			if ev.Name() == name {
+				return true
+			}
+		}
+	}
+}

--- a/pkg/util/test/events_test.go
+++ b/pkg/util/test/events_test.go
@@ -83,3 +83,34 @@ func TestCollectEvents(t *testing.T) {
 		testEvent2,
 	})
 }
+
+func TestRedirectEvents(t *testing.T) {
+	ctx := Context()
+	testEvent1 := events.Define("test-redirect-events-1", "test-event-1")(ctx, ttnpb.ApplicationIdentifiers{ApplicationID: "test-app"}, nil)
+	testEvent2 := events.Define("test-redirect-events-2", "test-event-2")(ctx, ttnpb.GatewayIdentifiers{GatewayID: "test-gtw"}, nil)
+	ch := make(chan events.Event, 2)
+	defer RedirectEvents(ch)()
+	events.Publish(testEvent1)
+	events.Publish(testEvent2)
+	close(ch)
+	var evs []events.Event
+	for ev := range ch {
+		evs = append(evs, ev)
+	}
+	assertions.New(t).So(
+		evs, should.Resemble, []events.Event{
+			testEvent1,
+			testEvent2,
+		},
+	)
+}
+
+func TestWaitEvent(t *testing.T) {
+	ctx := Context()
+	ch := make(chan events.Event, 1)
+	ch <- events.Define("test-wait-event-1", "test-event-1")(ctx, ttnpb.ApplicationIdentifiers{ApplicationID: "test-app"}, nil)
+	defer close(ch)
+	assertions.New(t).So(
+		WaitEvent(ctx, ch, "test-wait-event-1"),
+		should.BeTrue)
+}

